[Resolve #1252] Supporting resolvers in Hook and Resolver arguments, with new !substitute, !join, !split, and !select resolvers! (#1313)
Sceptre/sceptre: docs/_source/docs/resolvers.rst

.. _stack_attr_resolver:

select
~~~~~~

This resolver allows you to select a specific index of a list of items. This is great for combining
with the ``!split`` resolver to obtain part of a string. This function works almost the same as
CloudFormation's ``!Select`` intrinsic function, **except you can use this with negative indexes to

> select with a reverse index**.

Reviewer: What about "except you can use this with negative indices to select from the end of the list." ?

> CloudFormation's ``!Select`` intrinsic function, **except you can use this with negative indexes to
> select with a reverse index**.

Author: ```suggestion
CloudFormation's ``!Select`` intrinsic function, **except you can use this with negative indices to
select from the end of a list**.
```

The argument for this resolver should be a list with two elements: (1) A numerical index and (2) a
list of items to select out of. If the index is negative, it will select from the end of the list.
For example, "-1" would select the last element and "-2" would select the second-to-last element.

Example:

.. code-block:: yaml

   sceptre_user_data:
     # This selects the last element after you split the connection string on "/"
     DatabaseName: !select
       - -1
       - !split ["/", !stack_output my/database/stack.yaml::ConnectionString]

split
~~~~~

This resolver will split a value on a given delimiter string. This is great when combining with the
``!select`` resolver. This function works the same as CloudFormation's ``!Split`` intrinsic function.

Note: The return value of this resolver is a *list*, not a string. This will not work to set Stack
configurations that expect strings, but it WILL work to set Stack configurations that expect lists.

The argument for this resolver should be a list with two elements: (1) The delimiter to split on and
(2) a string to split.

Example:

.. code-block:: yaml

   notifications: !split
     - ";"

>      - !stack_output my/sns/topics.yaml::SemicolonDelimitedArns

Author: Actually, I don't really. !sub, !join, !select, and !split are ONLY useful when you have resolvers to string together; They're clumsy if you don't. It's better to use Jinja to do these things, if you can. But if you have resolvers whose resolved values you need to manipulate/interpolate/select from, they're really the only way to do that. 

As such, If we don't show how these resolvers can be used with other resolvers, we're not really showing how to use them well.

>      - !stack_output my/sns/topics.yaml::SemicolonDelimitedArns

Reviewer: ¿Por qué no los dos? :)

```yaml
   notifications: !split
     - ";"
     - "note1; note2; note3"
```

Resolves to:

```yaml
   notifications: [ "note1", "note2", "note3" ]
```

As a more practical example, the following will split a semicolon-delimited list exported by another stack:

```yaml
   notifications: !split
     - ";"
     - !stack_output my/sns/topics.yaml::SemicolonDelimitedArns
```

>      - !stack_output my/sns/topics.yaml::SemicolonDelimitedArns

Author: @dboitnot, sure, that _works_. But If you wanted to split a simple string, you'd just go like this:

```yaml
notifications: {{ "note1; note2; note3".split(";") }}
```

I fear that if we show people examples of using these resolvers in cases where you don't actually _need_ to use them (and where you really _shouldn't_ use them), it will be more confusing. The only times you need to use these resolvers is when you need to combine/manipulate the outputs of other resolvers. I actually added documentation to that effect to make it absolutely clear.

   parameters:
     VpcIdParameter: !stack_output_external prj-network-vpc::VpcIdOutput prod

> substitute

Reviewer: I don't have super-strong feelings about this, but why "substitute" and not "sub" as in CF? I might argue in favor of "sub" for consistency with CF.

Author: I don't know. I guess I just don't like the abbreviation. It doesn't seem as clear as `substitute`. @zaro0508 , thoughts on this?

Reviewer: I agree "substitute" has more intrinsic clarity, but since most of our users are going to be pretty familiar with CF and `sub` I think we could avoid the mental dichotomy by sticking to CF's nomenclature. But I would like @zaro0508's thoughts as well.

Author: I've changed this to `sub`

~~~~~~~~~~

This resolver allows you to create a string using Python string format syntax. This functions as a
great way to combine together a number of resolver outputs into a single string. This functions very
similarly to Cloudformation's ``!Sub`` intrinsic function.

The argument to this resolver should be a two-element list: (1) Is the format string, using
curly-brace templates to indicate variables, and (2) a dictionary where the keys are the format
string's variable names and the values are the variable values.

Example:

.. code-block:: yaml

   parameters:
     ConnectionString: !sub

>        - "postgres://{username}:{password}@{hostname}:{port}/{database}"

Author: `!sub` _needs_ a second argument that defines the variable values. It's not the same as CloudFormation's sub that lets you just resolve values without defining them. It's just a glorified `str.format()` invocation. It doesn't have a "scope" outside its own arguments. If you want to resolve something from the Stack scope, you could combine `!sub` and Jinja syntax together, like this:

```python
!sub
  # Note the double-curly braces is JINJA syntax to access variables within Jinja scope
  - "postgres://{{ var.username }}:{password}@{hostname}:{port}/{{ database }}"
  - password: !ssm /my/ssm/password
     hostname: !stack_output my/database/stack.yaml::HostName
     port: !stack_output my/database/stack.yaml::Port
```

Notice how, if it's accessible within Jinja, you don't actually need to provide that to `!sub` as a variable, because Jinja interpolation will happen BEFORE the yaml is rendered. I'm going to look into clarifying this in the docs, since I doubt you'll be the first one confused about this.
